File: ios/ColdOpenCore/Tests/ColdOpenCoreTests/QuoteSelectorTests.swift
import XCTest
@testable import ColdOpenCore

final class QuoteSelectorTests: XCTestCase {
    func testNeverRepeatsBeforeExhaustingPool() {
        let quotes = (0..<5).map { i in
            Quote(text: "Quote \(i)", episodeTitle: "Ep", season: 1, episode: i + 1, arc: "Arc")
        }
        let selector = QuoteSelector(quotes: quotes)

        var seen: Set<String> = []
        for _ in 0..<5 {
            guard let quote = selector.next() else {
                XCTFail("expected a quote")
                return
            }
            XCTAssertFalse(seen.contains(quote.text), "quote repeated before pool exhausted")
            seen.insert(quote.text)
        }
        XCTAssertEqual(seen.count, 5)
    }

    func testRefillsPoolAfterExhaustion() {
        let quotes = [Quote(text: "Only one", episodeTitle: "Ep", season: 1, episode: 1, arc: "Arc")]
        let selector = QuoteSelector(quotes: quotes)

        XCTAssertEqual(selector.next()?.text, "Only one")
        XCTAssertEqual(selector.next()?.text, "Only one")
    }

    func testEmptyCorpusReturnsNil() {
        let selector = QuoteSelector(quotes: [])
        XCTAssertNil(selector.next())
    }

    func testShownTextsAccumulatesAsQuotesAreDrawn() {
        let quotes = (0..<3).map { i in
            Quote(text: "Quote \(i)", episodeTitle: "Ep", season: 1, episode: i + 1, arc: "Arc")
        }
        let selector = QuoteSelector(quotes: quotes)

        XCTAssertEqual(selector.shownTexts, [])
        guard let first = selector.next() else { return XCTFail("expected a quote") }
        XCTAssertEqual(selector.shownTexts, [first.text])
        guard let second = selector.next() else { return XCTFail("expected a quote") }
        XCTAssertEqual(selector.shownTexts, [first.text, second.text])
    }

    func testResumingWithASavedShownSetExcludesThoseQuotesUntilThePoolExhausts() {
        let quotes = (0..<3).map { i in
            Quote(text: "Quote \(i)", episodeTitle: "Ep", season: 1, episode: i + 1, arc: "Arc")
        }
        // Simulate a WidgetKit extension relaunch after "Quote 0" and "Quote 1" were already shown.
        let selector = QuoteSelector(quotes: quotes, alreadyShown: ["Quote 0", "Quote 1"])

        XCTAssertEqual(selector.next()?.text, "Quote 2")
        // Pool is now exhausted (all 3 shown across the relaunch boundary) — refills.
        var seen: Set<String> = []
        for _ in 0..<3 {
            guard let quote = selector.next() else { return XCTFail("expected a quote") }
            seen.insert(quote.text)
        }
        XCTAssertEqual(seen, ["Quote 0", "Quote 1", "Quote 2"])
    }

    func testResumingWithEveryQuoteAlreadyShownRefillsImmediately() {
        let quotes = [Quote(text: "Only one", episodeTitle: "Ep", season: 1, episode: 1, arc: "Arc")]
        let selector = QuoteSelector(quotes: quotes, alreadyShown: ["Only one"])

        XCTAssertEqual(selector.next()?.text, "Only one")
    }

    func testRefillingAfterExhaustionClearsTheShownSetNotJustThePool() {
        let quotes = [Quote(text: "Only one", episodeTitle: "Ep", season: 1, episode: 1, arc: "Arc")]
        let selector = QuoteSelector(quotes: quotes)

        _ = selector.next()
        XCTAssertEqual(selector.shownTexts, ["Only one"])
        _ = selector.next() // triggers refill since the pool was already empty
        XCTAssertEqual(selector.shownTexts, ["Only one"])
    }
}

File: windows/WidgetProvider.Core.Tests/QuoteSelectorTests.cs
using ColdOpen.WidgetProvider.Core;
using Xunit;

namespace ColdOpen.WidgetProvider.Core.Tests;

public class QuoteSelectorTests
{
    [Fact]
    public void NeverRepeatsBeforeExhaustingThePool()
    {
        var quotes = Enumerable.Range(0, 5)
            .Select(i => new Quote($"Quote {i}", "Ep", 1, i + 1, "Arc"))
            .ToList();
        var selector = new QuoteSelector(quotes);

        var seen = new HashSet<string>();
        for (var i = 0; i < 5; i++)
        {
            var quote = selector.Next() ?? throw new InvalidOperationException("expected a quote");
            Assert.True(seen.Add(quote.Text), "quote repeated before pool exhausted");
        }
        Assert.Equal(5, seen.Count);
    }

    [Fact]
    public void RefillsPoolAfterExhaustion()
    {
        var quotes = new List<Quote> { new("Only one", "Ep", 1, 1, "Arc") };
        var selector = new QuoteSelector(quotes);

        Assert.Equal("Only one", selector.Next()?.Text);
        Assert.Equal("Only one", selector.Next()?.Text);
    }

    [Fact]
    public void EmptyCorpusReturnsNull()
    {
        var selector = new QuoteSelector(new List<Quote>());
        Assert.Null(selector.Next());
    }

    [Fact]
    public void ShownTextsAccumulatesAsQuotesAreDrawn()
    {
        var quotes = Enumerable.Range(0, 3)
            .Select(i => new Quote($"Quote {i}", "Ep", 1, i + 1, "Arc"))
            .ToList();
        var selector = new QuoteSelector(quotes);

        Assert.Empty(selector.ShownTexts);
        var first = selector.Next()!;
        Assert.Equal(new HashSet<string> { first.Text }, selector.ShownTexts);
        var second = selector.Next()!;
        Assert.Equal(new HashSet<string> { first.Text, second.Text }, selector.ShownTexts);
    }

    [Fact]
    public void ResumingWithASavedShownSetExcludesThoseQuotesUntilThePoolExhausts()
    {
        var quotes = Enumerable.Range(0, 3)
            .Select(i => new Quote($"Quote {i}", "Ep", 1, i + 1, "Arc"))
            .ToList();
        // Simulate a provider restart after "Quote 0" and "Quote 1" were already shown.
        var selector = new QuoteSelector(quotes, alreadyShown: new[] { "Quote 0", "Quote 1" });

        Assert.Equal("Quote 2", selector.Next()?.Text);
        // Pool is now exhausted (all 3 shown across the restart boundary) — refills.
        var seen = new HashSet<string>();
        for (var i = 0; i < 3; i++)
        {
            seen.Add(selector.Next()!.Text);
        }
        Assert.Equal(new HashSet<string> { "Quote 0", "Quote 1", "Quote 2" }, seen);
    }

    [Fact]
    public void ResumingWithEveryQuoteAlreadyShownRefillsImmediately()
    {
        var quotes = new List<Quote> { new("Only one", "Ep", 1, 1, "Arc") };
        var selector = new QuoteSelector(quotes, alreadyShown: new[] { "Only one" });

        Assert.Equal("Only one", selector.Next()?.Text);
    }

    [Fact]
    public void RefillingAfterExhaustionClearsTheShownSetNotJustThePool()
    {
        var quotes = new List<Quote> { new("Only one", "Ep", 1, 1, "Arc") };
        var selector = new QuoteSelector(quotes);

        selector.Next();
        Assert.Equal(new HashSet<string> { "Only one" }, selector.ShownTexts);
        selector.Next(); // triggers refill since the pool was already empty
        Assert.Equal(new HashSet<string> { "Only one" }, selector.ShownTexts);
    }
}
